### ISSUE TITLE
cgroups: unable to detect memory limits with systemd

### ISSUE THREAD

**Issue by antimatter15** [cockroachdb/cockroach#59236]
**Describe the problem**

I tried running CockroachDB from a `systemd` service with a specified `MemoryMax`.

**To Reproduce**

1. Install stock Ubuntu 20.04 LTS
2. Create systemd service 
```
[Service]
ExecStart=cockroach start-single-node --insecure --store /data
MemoryMax=100M
```
3. Run `systemctl start cockroach`
4. Observe `/data/log/cockroach.log` begins with 

```
available memory from cgroups is unsupported, using system memory 3.8 GiB instead: can't read available memory from cgroup v2 at /sys/fs/cgroup/unified/system.slice/cockroach.service/memory.max: open /sys/fs/cgroup/unified/system.slice/cockroach.service/memory.max: no such file or directory
server/status/recorder.go:605
```

**Expected behavior**
It should detect that the `MemoryMax` is 100M.

Changing to the legacy `MemoryLimit` doesn't do anything either. 

**Environment:**
 - CockroachDB version 20.2.3
 - Server OS: Ubuntu 20.04 LTS

**Additional context**
What was the impact?

CockroachDB often allocates too much memory and causes an OOM, rendering the system unusable. Note that the cgroups are still properly enforcing the out of memory limits— CockroachDB just isn't noticing them and then recklessly allocating so much memory as to exceed and get killed. 

-----

I've spent a few hours debugging this, and this is my understanding of what's going on and how to fix it:

By default most linux distros enable both v1 and v2 cgroups (https://wiki.archlinux.org/index.php/cgroups#As_root). 

However, mounting certain controllers (e.g. the memory controller) with cgroups v1 and cgroups v2 are mutually exclusive. So when cgroups v1 is enabled, it's not possible (at least with systemd) to configure cgroups v2 memory limits. 

When CockroachDB reads /proc/self/mountinfo (https://github.com/cockroachdb/cockroach/blob/master/pkg/util/cgroups/cgroups.go#L61), it'll often get something like this:

```
34 27 0:28 / /sys/fs/cgroup ro,nosuid,nodev,noexec shared:7 - tmpfs tmpfs ro,mode=755
35 34 0:29 / /sys/fs/cgroup/unified rw,nosuid,nodev,noexec,relatime shared:8 - cgroup2 cgroup2 rw,nsdelegate
36 34 0:30 / /sys/fs/cgroup/systemd rw,nosuid,nodev,noexec,relatime shared:9 - cgroup cgroup rw,xattr,name=systemd
37 27 0:31 / /sys/fs/pstore rw,nosuid,nodev,noexec,relatime shared:10 - pstore pstore rw
38 27 0:32 / /sys/fs/bpf rw,nosuid,nodev,noexec,relatime shared:11 - bpf none rw,mode=700
39 34 0:33 / /sys/fs/cgroup/net_cls,net_prio rw,nosuid,nodev,noexec,relatime shared:15 - cgroup cgroup rw,net_cls,net_prio
40 34 0:34 / /sys/fs/cgroup/perf_event rw,nosuid,nodev,noexec,relatime shared:16 - cgroup cgroup rw,perf_event
41 34 0:35 / /sys/fs/cgroup/cpu,cpuacct rw,nosuid,nodev,noexec,relatime shared:17 - cgroup cgroup rw,cpu,cpuacct
42 34 0:36 / /sys/fs/cgroup/pids rw,nosuid,nodev,noexec,relatime shared:18 - cgroup cgroup rw,pids
43 34 0:37 / /sys/fs/cgroup/rdma rw,nosuid,nodev,noexec,relatime shared:19 - cgroup cgroup rw,rdma
44 34 0:38 / /sys/fs/cgroup/blkio rw,nosuid,nodev,noexec,relatime shared:20 - cgroup cgroup rw,blkio
45 34 0:39 / /sys/fs/cgroup/hugetlb rw,nosuid,nodev,noexec,relatime shared:21 - cgroup cgroup rw,hugetlb
46 34 0:40 / /sys/fs/cgroup/cpuset rw,nosuid,nodev,noexec,relatime shared:22 - cgroup cgroup rw,cpuset
47 34 0:41 / /sys/fs/cgroup/freezer rw,nosuid,nodev,noexec,relatime shared:23 - cgroup cgroup rw,freezer
48 34 0:42 / /sys/fs/cgroup/devices rw,nosuid,nodev,noexec,relatime shared:24 - cgroup cgroup rw,devices
49 34 0:43 / /sys/fs/cgroup/memory rw,nosuid,nodev,noexec,relatime shared:25 - cgroup cgroup rw,memory
```

It scans through the list of mountpoints for a cgroup1 entry with the appropriate controller (i.e. /sys/fs/cgroup/memory) or any entry about cgroup2. In this case, the cgroup2 mount happens before the cgroup1 mount, so that gets read first. It checks to see if there's a `memory.max`, which there isn't— because it's not possible to add a cgroupv2 memory controller when cgroupv1 is also running. As soon as it hits this failure point it goes and declares that no cgroup limits could be found and falls back onto trying to use all system memory. 

**Proposed fix**

https://github.com/cockroachdb/cockroach/blob/master/pkg/util/cgroups/cgroups.go#L61

`getCgroupMem` should check for both cgroup1 and cgroup2 limits before falling back to using all the system's memory.
 

Jira issue: CRDB-6291

**Comment by knz** [comment]
@itsbilal I believe your changes have addressed this right? Feel free to close if they did.

**Comment by knz** [comment]
@abarganier is your recent PR a fix for this issue?

**Comment by abarganier** [comment]
@knz I don't think so (or at least it was not my intention) - I wasn't aware of this issue.

**Comment by srosenberg** [comment]
@antimatter15 Thanks for reporting this issue! Sorry it took nearly a year to address it :( The fix will be in the upcoming release of `22.2`. The nightly build is also available for _testing_.  E.g., let's set memory limit to `1G` and kick off via `systemd`,

```
wget -O cockroach https://storage.googleapis.com/cockroach-edge-artifacts-prod/cockroach/cockroach.linux-gnu-amd64.4329c6030405064eaed1c41c0237dd6c8b837b1a && chmod u+x cockroach

NUM_FILES_LIMIT=65536
MEMORY_MAX=1G
systemd-run --unit cockroach --same-dir --uid "$(id -u)" --gid "$(id -g)" --service-type=notify -p NotifyAccess=all -p "MemoryMax=${MEMORY_MAX}" -p LimitCORE=infinity -p "LimitNOFILE=${NUM_FILES_LIMIT}" ./cockroach start-single-node --insecure
```

Verify that the limit has been applied,

```
service cockroach status

Loaded: loaded (/run/systemd/transient/cockroach.service; transient)
  Transient: yes
     Active: active (running) since Thu 2022-11-17 03:35:12 UTC; 30s ago
   Main PID: 1902343 (cockroach)
      Tasks: 29 (limit: 38508)
     Memory: 165.1M (max: 1.0G)
     CGroup: /system.slice/cockroach.service
```
